### DISCUSSION TITLE
Lettuce OpenTelemetry Duration - What does it really measure?

### DISCUSSION THREAD

**Discussion by SimoneGiusso** [open-telemetry/opentelemetry-java-instrumentation#12117] [Q&A]
I didn't find any reference explaining exactly what the duration specified in the trace, really indicates. I'm trying to analyze some slow http requests, and I see some `MGET redis` requests taking even 750 ms to run. But what this duration includes? This is what  I suppose it includes:

1. Network Latency: Time for the request to travel from the client to the server and back.
2. Server Processing Time: Time taken by the server to process the request (in this case, the /assets/discover endpoint).
3. Database or Cache Access Time: If the operation involves querying a database or a cache (e.g., Redis), the time spent on these operations is also included.
4. Application Logic: Any time taken by the application in handling the request, like reading from the socket and deserialize the response.

I'm not sure about the last one. But what I like to know is, if I can exclude at least latencies coming from the app. So I can be sure the app is not the problem. Otherwise the trace is not helping much. Too many variables to consider.

**Comment by trask** [comment]
hi @SimoneGiusso, once we have profiling correlation then you could look at profiling stack traces to narrow down more precisely what was happening during those 750ms

you may also want to look at the underlying lettuce instrumentation, IIRC it's based on tracing callbacks that are coming from the underlying lettuce library

**Reply by SimoneGiusso** [reply]
What do you mean by profiling stack traces. In Jaeger I have this information displayed:

<img width="746" alt="Screenshot 2024-08-27 at 16 47 58" src="https://github.com/user-attachments/assets/68b36799-2753-4944-9f55-d65224335f6e">

_you may also want to look at the underlying lettuce instrumentation, IIRC it's based on tracing callbacks that are coming from the underlying lettuce library_

Also this statement is not clear to me. What IIRC stands for?

Thank you.

**Reply by SimoneGiusso** [reply]
Ping @trask 🙏

**Reply by laurit** [reply]
> What IIRC stands for?

if I remember correctly

>> once we have profiling correlation then you could look at profiling stack traces to narrow down more precisely what was happening during those 750ms

> What do you mean by profiling stack traces. In Jaeger I have this information displayed:

You can view this as a forward looking statement, once we have added profiling to the agent it might be easier to diagnose such issues. Meanwhile you have to use other means.

Have a look at https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/lettuce/lettuce-5.1/library/src/main/java/io/opentelemetry/instrumentation/lettuce/v5_1/OpenTelemetryTracing.java as you can see we rely on lettuce for the tracing. If you need to get a deeper understanding I'd suggest looking at lettuce source code or asking the lettuce community for the details on when exactly these methods are called.

**Comment by SimoneGiusso** [comment]
> once we have added profiling to the agent it might be easier to diagnose such issues.

I went though these articles:
- [OpenTelemetry announces support for profiling](https://opentelemetry.io/blog/2024/profiling/)
- [Elastic Contributes its Continuous Profiling Agent to OpenTelemetry
](https://opentelemetry.io/blog/2024/elastic-contributes-continuous-profiling-agent/)

I'm really looking forward to this 🙂! Thank you!

I close the discussion, and investigate further with lettuce community if necessary.

Thanks both of you for your time.
